pkg/instrumentation/instrumentation.go

```diff
@@ -15,70 +15,63 @@
 package instrumentation
 
 import (
-	"strings"
-	"time"
+	"fmt"
+	"net/http"
 
-	"contrib.go.opencensus.io/exporter/jaeger"
-	"contrib.go.opencensus.io/exporter/prometheus"
-	prom "github.com/prometheus/client_golang/prometheus"
-	model "github.com/prometheus/client_model/go"
 	"go.opencensus.io/plugin/ocgrpc"
-	"go.opencensus.io/stats/view"
-	"go.opencensus.io/trace"
 	"google.golang.org/grpc"
 
 	logger "github.com/intel/cri-resource-manager/pkg/log"
 )
 
 const (
-	// ServiceName is our Jaeger and Prometheus service names.
+	// ServiceName is our service name in external tracing and metrics services.
 	ServiceName = "CRI-RM"
 )
 
 // Our logger instance.
-var log logger.Logger = logger.NewLogger("tracing")
+var log logger.Logger = logger.NewLogger("instrumentation")
 
-// Our Jaeger trace and Prometheus metrics exporters.
-var jexport *jaeger.Exporter
-var pexport *prometheus.Exporter
+// Our instrumentation service instance.
+var service = createService()
 
-// prometheus Gatherers dynamically registered with us.
-var dynamicGatherers = &gatherers{gatherers: prom.Gatherers{}}
+// Get drop-in mux for external services.
+func GetHTTPMux() *http.ServeMux {
+	return service.ServeMux
+}
 
-// ConfigureTracing configures Jaeger-tracing.
-func ConfigureTracing(tc TraceConfig) error {
-	log.Debug("applying trace configuration %v", tc)
-	trace.ApplyConfig(trace.Config{DefaultSampler: tc.Sampler()})
-	return nil
+// ConfigureTracing configures sampling for Jaeger tracing.
+func ConfigureTracing(c TraceConfig) {
+	service.ConfigureTracing(c)
 }
 
-// TracingEnabled returns true if the tracing sampler is not disabled.
+// TracingEnabled returns true if the Jaeger tracing sampler is not disabled.
 func TracingEnabled() bool {
-	return float64(opt.Trace) > 0.0
+	return service.TracingEnabled()
 }
 
-// Start sets up instrumentation (tracing, metrics collection, etc.).
+// Start our internal instrumentation services.
 func Start() error {
-	log.Info("Starting instrumenation...")
-	ConfigureTracing(opt.Trace)
-
-	registerJaegerExporter()
-	registerGrpcTraceViews()
-	registerPrometheusExporter()
-
-	if err := HTTPStart(); err != nil {
-		return traceError("failed to start tracing/metrics HTTP server: %v", err)
+	if service == nil {
+		log.Error("cannot start, instrumentation service creation failed")
 	}
-
-	return nil
+	return service.Start()
 }
 
-// Stop shuts down instrumentation.
+// Stop stops our internal instrumentation services.
 func Stop() {
-	unregisterJaegerExporter()
-	unregisterGrpcTraceViews()
-	unregisterPrometheusExporter()
-	HTTPShutdown()
+	if service != nil {
+		service.Stop()
```

Comment:
should we log the possible error, just in case?

Reply:
Changed error handling to be more consistent. Now
  - stopping a non-existent service is not an error
  - starting/restarting a non-existent service is an error
  - the caller is expected to deal with/log the top-level errors
Fixed a couple of potential crashes for service == nil (Restart(), GetHTTPMux(), ...).

instrumentation.s